File: tests/refs/valid_anagram.cpp
// Reference C++ driver for tests/cases/valid_anagram.json.
#include <array>
#include <iostream>
#include <string>

bool isAnagram(const std::string& a, const std::string& b) {
    if (a.size() != b.size()) return false;
    std::array<int, 256> cnt{};
    for (unsigned char c : a) ++cnt[c];
    for (unsigned char c : b) if (--cnt[c] < 0) return false;
    return true;
}

static std::string parseStr(const std::string& line) {
    std::size_t sp = line.find(' ', 4);
    if (sp == std::string::npos) return std::string();
    int len = std::stoi(line.substr(4, sp - 4));
    std::string s = line.substr(sp + 1);
    if ((int)s.size() < len) s += std::string(len - s.size(), ' ');
    else if ((int)s.size() > len) s = s.substr(0, len);
    return s;
}

int main() {
    std::string line, name, a, b; int expected = 0;
    int strIdx = 0; bool he = false; int p = 0, f = 0;
    auto run = [&]() {
        if (name.empty() || strIdx < 2 || !he) return;
        bool got = isAnagram(a, b);
        if ((int)got == expected) { std::cout << "PASS valid_anagram :: " << name << "\n"; ++p; }
        else { std::cout << "FAIL valid_anagram :: " << name << "\n  expected: " << expected << "\n  got: " << got << "\n"; ++f; }
        name.clear(); a.clear(); b.clear(); strIdx = 0; he = false;
    };
    while (std::getline(std::cin, line)) {
        if (line.empty()) { run(); continue; }
        if (line.rfind("CASE ", 0) == 0) { name = line.substr(5); strIdx = 0; }
        else if (line.rfind("STR ", 0) == 0) {
            std::string s = parseStr(line);
            if (strIdx == 0) { a = s; ++strIdx; } else { b = s; ++strIdx; }
        } else if (line.rfind("BOOL ", 0) == 0) { expected = std::stoi(line.substr(5)); he = true; }
    }
    run();
    std::cout << "TOTAL: " << p << " passed, " << f << " failed\n";
    return f == 0 ? 0 : 1;
}

File: tests/refs/valid_anagram.rs
// Reference Rust driver for tests/cases/valid_anagram.json.
use std::io::{self, BufRead};
use std::collections::HashMap;

fn is_anagram(a: &str, b: &str) -> bool {
    if a.chars().count() != b.chars().count() { return false; }
    let mut cnt: HashMap<char, i32> = HashMap::new();
    for c in a.chars() { *cnt.entry(c).or_insert(0) += 1; }
    for c in b.chars() {
        let e = cnt.entry(c).or_insert(0); *e -= 1;
        if *e < 0 { return false; }
    }
    true
}

fn parse_str(line: &str) -> String {
    let rest = &line[4..];
    if let Some(sp) = rest.find(' ') {
        let len: usize = rest[..sp].parse().unwrap_or(0);
        let s: String = rest[sp+1..].to_string();
        if s.chars().count() < len { let mut s = s; s.push_str(&" ".repeat(len - s.chars().count())); s }
        else { s }
    } else { String::new() }
}

fn main() {
    let stdin = io::stdin();
    let mut name = String::new(); let mut a = String::new(); let mut b = String::new();
    let mut expected = 0i32;
    let mut str_idx = 0i32; let mut he = false;
    let mut p = 0u32; let mut f = 0u32;
    let flush = |name: &mut String, a: &mut String, b: &mut String, expected: i32,
                 str_idx: &mut i32, he: &mut bool, p: &mut u32, f: &mut u32| {
        if !name.is_empty() && *str_idx >= 2 && *he {
            let got = is_anagram(a, b);
            if (got as i32) == expected { println!("PASS valid_anagram :: {}", name); *p += 1; }
            else { println!("FAIL valid_anagram :: {}\n  expected: {}\n  got: {}", name, expected, got); *f += 1; }
        }
        name.clear(); a.clear(); b.clear(); *str_idx = 0; *he = false;
    };
    for line in stdin.lock().lines() {
        let line = line.unwrap_or_default();
        if line.is_empty() { flush(&mut name, &mut a, &mut b, expected, &mut str_idx, &mut he, &mut p, &mut f); continue; }
        if let Some(rest) = line.strip_prefix("CASE ") { name = rest.to_string(); str_idx = 0; }
        else if line.starts_with("STR ") {
            let s = parse_str(&line);
            if str_idx == 0 { a = s; str_idx += 1; } else { b = s; str_idx += 1; }
        }
        else if let Some(rest) = line.strip_prefix("BOOL ") { expected = rest.parse().unwrap_or(0); he = true; }
    }
    flush(&mut name, &mut a, &mut b, expected, &mut str_idx, &mut he, &mut p, &mut f);
    println!("TOTAL: {} passed, {} failed", p, f);
    std::process::exit(if f == 0 { 0 } else { 1 });
}
